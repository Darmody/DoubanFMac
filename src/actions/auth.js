// @flow
import * as types from 'constants/types/ActionTypes'
import type { Action } from 'constants/types/Redux'

export const login = (username: string, password: string): Action => ({
  type: types.LOGIN_REQUEST,
  payload: {
    username,
    password,
  }
})

export const logout = (): Action => ({
  type: types.LOGOUT,
  payload: undefined,
})

export const logined = (response: Object): Action => ({
  type: types.LOGIN_SUCCESS,
  payload: response,
})
