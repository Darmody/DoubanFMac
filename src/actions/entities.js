// @flow
import * as types from 'constants/types/ActionTypes'
import type { Action } from 'constants/types/Redux'

export const save = (data: Object): Action => ({
  type: types.ENTITIES_SAVE,
  payload: data,
})

export const update = (data: Object): Action => ({
  type: types.ENTITIES_UPDATE,
  payload: data,
})
