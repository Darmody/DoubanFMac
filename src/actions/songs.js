// @flow
import * as types from 'constants/types/ActionTypes'
import type { Action } from 'constants/types/Redux'

export const listen = (channel: number): Action => ({
  type: types.SONG_LISTEN_REQUEST,
  payload: channel,
})

export const next = (channel: number, sid?: number): Action => ({
  type: types.SONG_NEXT_REQUEST,
  payload: {
    channel,
    sid,
  }
})

export const mark = (channel: number, sid: number): Action => ({
  type: types.SONG_MARK_REQUEST,
  payload: {
    sid,
    channel,
  }
})

export const current = (song: {}): Action => ({
  type: types.SONG_CURRENT,
  payload: song,
})
