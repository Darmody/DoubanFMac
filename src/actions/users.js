// @flow
import * as types from 'constants/types/ActionTypes'
import type { Action } from 'constants/types/Redux'

export const current = (): Action => ({
  type: types.USER_CURRENT_REQUEST,
  payload: undefined,
})
