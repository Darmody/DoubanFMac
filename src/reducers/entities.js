// @flow
import R from 'ramda'
import Immutable from 'seamless-immutable'
import * as types from 'constants/types/ActionTypes'
import type { Reducer } from 'constants/types/Redux'

const initialState = Immutable({
})

const reducer: Reducer = (state = initialState, { type, payload = {} }) => {
  switch (type) {
    case types.ENTITIES_SAVE: {
      let finalState = state
      R.forEachObjIndexed((data, model) =>
        R.forEachObjIndexed((value, id) => {
          finalState = Immutable.setIn(finalState, [model, id], value)
        })(data || {})
      )(payload)
      return finalState
    }
    case types.ENTITIES_UPDATE: {
      let finalState = state
      const updateField = ({ field, value }) => {
        finalState = Immutable.setIn(state, field, value)
      }

      R.forEach(updateField, payload)

      return finalState
    }
    default: return state
  }
}

export default reducer
