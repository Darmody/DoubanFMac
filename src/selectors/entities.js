// @flow
import { schema as Schema, denormalize } from 'normalizr'
import type { StateShape } from 'constants/types/Redux'

const isModelExists = (entities, model) => entities[model]

export const selectEntities = (state: StateShape, schema: Schema, ids: [] = []) => {
  const model = schema[0].key

  if (isModelExists(state.entities, model)) {
    return denormalize({ [model]: ids }, { [model]: schema }, state.entities)[model]
  }

  return []
}

export const selectEntity = (state: StateShape, schema: Schema, id: number = 0) => {
  const model = schema.key

  if (isModelExists(state.entities, model)) {
    return denormalize({ [model]: id }, { [model]: schema }, state.entities)[model]
  }

  return {}
}
