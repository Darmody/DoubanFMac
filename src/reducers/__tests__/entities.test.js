import Immutable from 'seamless-immutable'
import * as actions from 'actions/entities'
import reducer from 'reducers/entities'

describe('reducer::entities', () => {
  it('initialize state', () => {
    expect(reducer(undefined, {})).toMatchSnapshot()
  })

  it('存储 entity', () => {
    expect(
      reducer(undefined, actions.save({ users: { 1: { id: 1 } } })),
    ).toMatchSnapshot()
  })

  it('更新 entity 字段', () => {
    const newName = 'new name'
    const state = reducer(
      Immutable({ users: { 1: { name: 'name' } } }),
      actions.update([{ field: ['users', 1, 'name'], value: newName }])
    )

    expect(state).toMatchSnapshot()
    expect(state.users['1'].name).toBe(newName)
  })
})
