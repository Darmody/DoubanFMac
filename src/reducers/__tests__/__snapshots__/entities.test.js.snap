// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer::entities initialize state 1`] = `Object {}`;

exports[`reducer::entities 存储 entity 1`] = `
Object {
  "users": Object {
    "1": Object {
      "id": 1,
    },
  },
}
`;

exports[`reducer::entities 更新 entity 字段 1`] = `
Object {
  "users": Object {
    "1": Object {
      "name": "new name",
    },
  },
}
`;
